procrastinate/worker.py

```diff
@@ -289,74 +194,154 @@ async def run_job(self, job: jobs.Job, worker_id: int) -> None:
             )
-            if task_result:
-                text += f" - Result: {task_result}"[:250]
-            self.logger.log(log_level, text, extra=extra, exc_info=exc_info)
 
-    def stop(self):
-        # Ensure worker will stop after finishing their task
-        self.stop_requested = True
-        # Ensure workers currently waiting are awakened
-        if self.notify_event:
-            self.notify_event.set()
+            self.logger.debug(
+                f"Acknowledged job completion {job.call_string}",
+                extra=self.base_context.log_extra(action="finish_task", status=status),
+            )
 
-        # Logging
+    async def run(self):
+        self._run_task = asyncio.current_task()
+        notify_event = asyncio.Event()
 
         self.logger.info(
-            "Stop requested",
-            extra=self.base_context.log_extra(action="stopping_worker"),
+            f"Starting worker on {self.base_context.queues_display}",
+            extra=self.base_context.log_extra(
+                action="start_worker", queues=self.queues
+            ),
         )
 
-        contexts = [
-            context for context in self.current_contexts.values() if context.job
-        ]
-        now = time.time()
-        for context in contexts:
+        job_semaphore = asyncio.Semaphore(self.concurrency)
+
+        running_jobs: dict[JobContext, asyncio.Task] = {}
+        side_tasks = [asyncio.create_task(self.periodic_deferrer())]
+        if self.wait and self.listen_notify:
+            listener_coro = self.app.job_manager.listen_for_jobs(
+                event=notify_event,
+                queues=self.queues,
+            )
+            side_tasks.append(asyncio.create_task(listener_coro, name="listener"))
+
+        try:
+            context = contextlib.nullcontext()
+            if self.install_signal_handlers:
+                context = signals.on_stop(self.stop)
+            with context:
+                """Processes jobs until cancelled or until there is no more available job (wait=False)"""
+                while True:
+                    out_of_job = None
+                    while not out_of_job:
+                        # acquire job_semaphore so that a new job is not fetched if maximum concurrency is reached
+                        async with job_semaphore:
+                            job = await self.app.job_manager.fetch_job(
+                                queues=self.queues
+                            )
+                        if job:
+                            # job_semaphore should be acquired straight because it is
+                            # only acquired when not at full capacity at this time
+                            # however, shield it from cancellation in the unlikely event the worker
+                            # is cancelled at that precise time to not abandon the job
+                            await asyncio.shield(job_semaphore.acquire())
```

Comment:
I'm not sure I understand the reasoning here.
we acquire the semaphore, then get a job, then release the semaphore, then re-acquire the semaphore ?

Why don't we acquire it once before fetching a job, and release it once, when the job is done ?

Reply:
I guess we could do that too. 
This will also need to be released if no job is found and if fetch job was to raise. Let me try that and see if that reads better.